aws/rust-runtime/aws-inlineable/src/rds_auth_token.rs

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Code related to creating signed URLs for logging in to RDS.
+//!
+//! For more information, see <https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.html>
+
+use aws_credential_types::provider::{ProvideCredentials, SharedCredentialsProvider};
+use aws_sigv4::http_request;
+use aws_sigv4::http_request::{SignableBody, SignableRequest, SigningSettings};
+use aws_sigv4::sign::v4;
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::identity::Identity;
+use aws_types::region::Region;
+use std::fmt;
+use std::fmt::Debug;
+use std::ops::Deref;
+use std::time::Duration;
+
+const ACTION: &str = "connect";
+const SERVICE: &str = "rds-db";
+
+/// A signer that generates an auth token for a database.
+///
+/// ## Example
+///
+/// ```ignore
+/// use crate::auth_token::{AuthTokenGenerator, Config};
+///
+/// #[tokio::main]
+/// async fn main() {
+///    let cfg = aws_config::load_defaults(BehaviorVersion::latest()).await;
+///    let generator = AuthTokenGenerator::new(
+///        Config::builder()
+///            .hostname("zhessler-test-db.cp7a4mblr2ig.us-east-1.rds.amazonaws.com")
+///            .port(5432)
+///            .username("zhessler")
+///            .build()
+///            .expect("cfg is valid"),
+///    );
+///    let token = generator.auth_token(&cfg).await.unwrap();
+///    println!("{token}");
+/// }
+/// ```
+#[derive(Debug)]
+pub struct AuthTokenGenerator {
+    config: Config,
+}
+
+/// An auth token usable as a password for an RDS database.
+///
+/// This struct can be converted into a `&str` using the `Deref` trait or by calling `to_string()`.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AuthToken {
+    inner: String,
+}
+
+impl Deref for AuthToken {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
```

Comment:
I'm thinking whether we need to implement `Deref` _NOW_ since customers can get `&str` anyway from the `Display` trait. Implementing `Deref` feels a bit exposing the inner type and we cannot change the `Target` type in the future since it's part of public API.

Essentially, implementing `Deref` now seems to be a one-way door. Not implementing `Deref` now gives us a two-way door and cutstomers can still get `&str` by the `Display` trait.

Reply:
I agree `Deref` is a one way door but using `Display` trait is kind of an abuse/irregular use of that trait since customers won't be using it for display/formatting purposes. IMO I'd just offer an `as_str(): &str` method. If we ever implement deref it can forward to that 🤷

Reply:
Ah, actually [we get `to_string` for free](https://doc.rust-lang.org/std/string/trait.ToString.html#tymethod.to_string) once we've implemented the `Display` trait.